engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java

```diff
@@ -226,10 +310,12 @@ public void copyAsync(DataObject srcData, DataObject destData, Host destHost, As
 
                 if (canHandleSrc) {
                     String errMsg = "This operation is not supported (DataStoreCapabilities.STORAGE_SYSTEM_SNAPSHOT " +
-                            "not supported by source storage plug-in). " + getSrcDataStoreMsg(srcData);
+                            "not supported by source storage plug-in). " + getDestDataStoreMsg(destData);
```

Comment:
I can add a comment. Basically, there are three if statements in a row. The first one covers 'if (canHandleSrc && canHandleDest)'. The second one covers 'if (canHandleDest)' (so, at this point, you know that both canHandleSrc && canHandleDest can't resolve to true). The final one is 'if (canHandleSrc)' (so, we are saying that our storage plug-in can handle the source side, but not the destination side, which is a problem).

Reply:
This is a funny cause. I get amused imagining the thought process of a person that code these things...
Documentation could help. However, I would first suggesting extracting lines 191-227 to a method. Then, instead of starting with `if (canHandleSrc && canHandleDest)`, you can do the following:
if (!canHandleSrc){Throw exception regarding source not being able to handle the request}
if (!canHandleDest){Throw exception regarding destination not being able to handle the request }

Then, if you pass those two IFs, you do not need a third one to check `if (canHandleSrc && canHandleDest)`